core/src/main/scala/org/apache/spark/status/protobuf/TaskDataWrapperSerializer.scala

```diff
@@ -0,0 +1,138 @@
+      .setStageId(input.stageId)
+      .setStageAttemptId(input.stageAttemptId)
+    input.errorMessage.foreach(builder.setErrorMessage)
+    input.accumulatorUpdates.foreach { update =>
+      builder.addAccumulatorUpdates(serializeAccumulableInfo(update))
+    }
+    builder.build().toByteArray
+  }
+
+  def deserialize(bytes: Array[Byte]): TaskDataWrapper = {
+    val binary = StoreTypes.TaskDataWrapper.parseFrom(bytes)
+    val accumulatorUpdates = new ArrayBuffer[AccumulableInfo]()
+    binary.getAccumulatorUpdatesList.forEach { update =>
+      accumulatorUpdates.append(new AccumulableInfo(
+        id = update.getId,
+        name = update.getName,
+        update = getOptional(update.hasUpdate, update.getUpdate),
+        value = update.getValue))
+    }
+    new TaskDataWrapper(
+      taskId = binary.getTaskId,
+      index = binary.getIndex,
+      attempt = binary.getAttempt,
+      partitionId = binary.getPartitionId,
+      launchTime = binary.getLaunchTime,
+      resultFetchStart = binary.getResultFetchStart,
+      duration = binary.getDuration,
+      executorId = binary.getExecutorId,
+      host = binary.getHost,
+      status = binary.getStatus,
+      taskLocality = binary.getTaskLocality,
+      speculative = binary.getSpeculative,
+      accumulatorUpdates = accumulatorUpdates.toSeq,
+      errorMessage = getOptional(binary.hasErrorMessage, binary.getErrorMessage),
+      hasMetrics = binary.getHasMetrics,
+      executorDeserializeTime = binary.getExecutorDeserializeTime,
+      executorDeserializeCpuTime = binary.getExecutorDeserializeCpuTime,
+      executorRunTime = binary.getExecutorRunTime,
+      executorCpuTime = binary.getExecutorCpuTime,
+      resultSize = binary.getResultSize,
+      jvmGcTime = binary.getJvmGcTime,
+      resultSerializationTime = binary.getResultSerializationTime,
+      memoryBytesSpilled = binary.getMemoryBytesSpilled,
+      diskBytesSpilled = binary.getDiskBytesSpilled,
+      peakExecutionMemory = binary.getPeakExecutionMemory,
+      inputBytesRead = binary.getInputBytesRead,
+      inputRecordsRead = binary.getInputRecordsRead,
+      outputBytesWritten = binary.getOutputBytesWritten,
+      outputRecordsWritten = binary.getOutputRecordsWritten,
+      shuffleRemoteBlocksFetched = binary.getShuffleRemoteBlocksFetched,
+      shuffleLocalBlocksFetched = binary.getShuffleLocalBlocksFetched,
+      shuffleFetchWaitTime = binary.getShuffleFetchWaitTime,
+      shuffleRemoteBytesRead = binary.getShuffleRemoteBytesRead,
+      shuffleRemoteBytesReadToDisk = binary.getShuffleRemoteBytesReadToDisk,
+      shuffleLocalBytesRead = binary.getShuffleLocalBytesRead,
+      shuffleRecordsRead = binary.getShuffleRecordsRead,
+      shuffleBytesWritten = binary.getShuffleBytesWritten,
+      shuffleWriteTime = binary.getShuffleWriteTime,
+      shuffleRecordsWritten = binary.getShuffleRecordsWritten,
+      stageId = binary.getStageId.toInt,
+      stageAttemptId = binary.getStageAttemptId
```

Comment:
Discuss: I am yet to look through all the codepaths, but I am wondering if use of `weakIntern` in `LiveEntity` is relevant to these changes.

Reply:
I think that's a bit complicated. I prefer keeping it simple for now. The project helps reduce memory usage already.

Reply:
That fix was added based on what was observed in production history server - where the gc impact and the number of duplication objects was extremely high.

Reply:
Could you provide more details? Do you mean using `weakIntern` for all the string fields?  Will it help the GC if we use `weakIntern(binary.getHost)`?

Reply:
You can take a look at `LiveEntity` on which fields it helps with - this was based on what we saw inhouse.
In case of history server, the number of hosts in a cluster is fairly limited (depends on size of cluster) - but the number of applications + tasks being very large.

Reply:
@mridulm Thanks. I have updated the code according to https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/status/LiveEntity.scala#L208